crypto/cms/cms_env.c

```diff
@@ -122,6 +124,39 @@ CMS_ContentInfo *CMS_EnvelopedData_create(const EVP_CIPHER *cipher)
     return NULL;
 }
 
+int cms_EnvelopedData_final(CMS_ContentInfo *cms, BIO *chain)
+{
+    CMS_EnvelopedData *env = NULL;
+    EVP_CIPHER_CTX *ctx = NULL;
+    BIO *mbio = BIO_find_type(chain, BIO_TYPE_CIPHER);
+
+    env = cms_get0_enveloped(cms);
+    if (!env)
+        return 0;
+
+    if (!mbio) {
+        CMSerr(CMS_F_CMS_ENVELOPEDDATA_FINAL, CMS_R_CONTENT_NOT_FOUND);
+        return 0;
+    }
+
+    BIO_get_cipher_ctx(mbio, &ctx);
+
+    /*
+     * If the selected cipher supports unprotected attributes,
+     * deal with it using special ctrl function
+     */
+    if (EVP_CIPHER_flags(EVP_CIPHER_CTX_cipher(ctx)) & EVP_CIPH_FLAG_CIPHER_WITH_MAC) {
+      cms->d.envelopedData->unprotectedAttrs = sk_X509_ATTRIBUTE_new_null();
+      if (EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_PROCESS_UNPROTECTED, 1, env->unprotectedAttrs) <= 0) {
```

Comment:
?? - Is this like a callback to allow attributes to be tacked on by the application.
If it is it might be nice to document it somewhere....
There must be a nicer way of doing this (even if there is a callback - maybe it should have a API to add attributes).

Reply:
I like the idea of documenting it though don't understand yet where to place this documentation. 
The intention here is not application-level attributes but cipher-level ones.

Reply:
My point is that this is a list of attributes which the user could potentially add their own. It looks like this code is hijacking the attributes for its own purposes and then potentially leaking the user attributes (if they are ever added).

Reply:
Thanks for the clarification!

Sorry, I don't understand why do you treat passing the attributes (in fact, they were always absent here) to a ctrl-function as a leak?

I need to add some alg-specific attributes to an (empty) stack. How can I do it not passing the stack?

Reply:
A better pattern might be to always allocate the stack so that it is normally has an empty one. Then your control can just add to this list (so if a user does need to add attributes they can).

Reply:
Always allocating the stack causes a tests failure. But I've added the check whether it's already present.